art/_install.py

```diff
@@ -0,0 +1,104 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import absolute_import
+
+import os
+import shutil
+import stat
+import click
+
+from . import _cache
+from . import _paths
+
+class InstallUnmatchedError(click.ClickException):
+    error = '''Source path(s) did not match any files in the archive:
+  archive: {archive}
+  project: {project}
+  job: {job}
+  ref: {ref}
+    {unmatched}'''
+
+    def __init__(self, filename, entry, unmatched):
+        unmatched_str = '\n    '.join((str(action) for action in unmatched))
+
+        message = self.error.format(
+            archive=_cache.cache_path(filename),
+            unmatched=unmatched_str,
+            **entry)
+
+        super().__init__(message)
+
+class InstallAction():
+    S_IRWXUGO = 0o0777
+
+    """Represents a user request to install a file from an artifact archive
+    """
+    def __init__(self, source, destination):
+        self.src = source
+        self.dest = destination
+        self.matched = False
+
+        if source == '.':
+            # "copy all" filter
+            self._match = lambda f: True
+            self.translate = lambda f: os.path.join(self.dest, f)
+        elif source.endswith('/'):
+            # 1:1 directory filter
+            self._match = lambda f: f.startswith(self.src)
+            self.translate = lambda f: os.path.join(self.dest, f[len(self.src):])
+        else:
+            # 1:1 file filter
+            self._match = lambda f: f == self.src
+            self.translate = lambda f: self.dest
+
+    def match(self, filepath):
+        """Compare an archive file using the source file pattern
+        Arguments:
+            filepath    The archive filepath to evaluate
+
+        Returns: True if the pattern matched the filepath, otherwise False
+        """
+        if not self._match(filepath):
+            return False
+
+        self.matched = True
```

Comment:
This looks weird... because it is. The `matched` attribute latches to `True` on any match. It needs a better name.

The `install` command calls `match` many times. It then uses `matched` to identify actions that didn't match any file in the archive.

Reply:
I decided to remove the "matched" state from the `InstallAction` class. The `install` command now uses the yaml dictionary to track artifact requests that were not found.